File: Level_4(C++,Java)/C++/Pen&Paper/Paper.cpp
#include "Paper.h"

Paper::Paper(int maxSymbols) : maxSymbols(maxSymbols), symbols(0) {}

Paper::~Paper() {}

int Paper::getMaxSymbols() const {
    return this->maxSymbols;
}

int Paper::getSymbols() const {
    return this->symbols;
}

void Paper::addContent(const std::string& message) {
    int freeSpace = this->maxSymbols - this->symbols;
    
    if ( this->symbols == this->maxSymbols ) {
        throw OutOfSpace();
    }
    
    this->content += message.substr(0, freeSpace);
    
    this->symbols = this->content.size();
}

void Paper::show() const {
    std::cout << "(" << this->symbols << '/' << this->maxSymbols << ')' 
              << "  Text: " << '[' << this->content << ']' << std::endl;
}

File: Level_4(C++,Java)/java/Pen&Paper/Paper.java
public class Paper {
    private int maxSymbols;
    private int symbols;
    private StringBuffer content;
    
    public Paper(int maxSymbols) {
        this.maxSymbols = maxSymbols;
        this.symbols = 0;
        content = new StringBuffer();
    }
    
    public Paper() {
        this(4096);
    }
    
    public int getMaxSymbols() {
        return this.maxSymbols;
    }
    
    public int getSymbols() {
        return this.symbols;
    }
    
    public void addContent(String message) throws OutOfSpaceException {
        int freeSpace = this.maxSymbols - this.symbols;
    
        if ( this.symbols == this.maxSymbols ) {
            throw new OutOfSpaceException();
        }
        
        if ( message.length() <= freeSpace ) {
            this.content.append(message);
        } else {
            this.content.append(message.substring(0, freeSpace));
        }
        
        this.symbols = this.content.length();
    }
    
    public void show() {
        System.out.print("(" + this.symbols + "/" + this.maxSymbols + ")"); 
        System.out.println("  Text: [" + this.content + "]");
    }
}
